File: skills/code-architect/references/code-examples/csharp/AbstractFactory.cs
// Abstract Factory — create families of related objects without specifying concrete classes.
using System;

namespace Patterns.AbstractFactory
{
    interface IElementX { string Tag(); }
    interface IElementY { string Tag(); }

    interface IFamilyFactory
    {
        IElementX CreateX();
        IElementY CreateY();
    }

    // Family 1
    sealed class Family1X : IElementX { public string Tag() => "X[Family1]"; }
    sealed class Family1Y : IElementY { public string Tag() => "Y[Family1]"; }
    sealed class Family1Factory : IFamilyFactory
    {
        public IElementX CreateX() => new Family1X();
        public IElementY CreateY() => new Family1Y();
    }

    // Family 2
    sealed class Family2X : IElementX { public string Tag() => "X[Family2]"; }
    sealed class Family2Y : IElementY { public string Tag() => "Y[Family2]"; }
    sealed class Family2Factory : IFamilyFactory
    {
        public IElementX CreateX() => new Family2X();
        public IElementY CreateY() => new Family2Y();
    }

    // Client code depends only on the abstract factory and product interfaces.
    sealed class Client
    {
        private readonly IElementX _x;
        private readonly IElementY _y;
        public Client(IFamilyFactory factory)
        {
            _x = factory.CreateX();
            _y = factory.CreateY();
        }
        public string Combine() => $"{_x.Tag()} + {_y.Tag()}";
    }

    static class Demo
    {
        public static void Run()
        {
            Console.WriteLine(new Client(new Family1Factory()).Combine());
            Console.WriteLine(new Client(new Family2Factory()).Combine());
        }

        public static void Main() => Run();
    }
}

File: skills/code-architect/references/code-examples/java/AbstractFactory.java
// Abstract Factory — create families of related objects without specifying concrete classes.

interface Button { String render(); }
interface Checkbox { String render(); }

class LightButton implements Button {
    public String render() { return "[ Light Button ]"; }
}
class LightCheckbox implements Checkbox {
    public String render() { return "( Light Checkbox )"; }
}
class DarkButton implements Button {
    public String render() { return "[[ Dark Button ]]"; }
}
class DarkCheckbox implements Checkbox {
    public String render() { return "(( Dark Checkbox ))"; }
}

interface WidgetFactory {
    Button createButton();
    Checkbox createCheckbox();
}

class LightWidgetFactory implements WidgetFactory {
    public Button createButton() { return new LightButton(); }
    public Checkbox createCheckbox() { return new LightCheckbox(); }
}

class DarkWidgetFactory implements WidgetFactory {
    public Button createButton() { return new DarkButton(); }
    public Checkbox createCheckbox() { return new DarkCheckbox(); }
}

class Screen {
    private final Button button;
    private final Checkbox checkbox;
    Screen(WidgetFactory factory) {
        this.button = factory.createButton();
        this.checkbox = factory.createCheckbox();
    }
    String draw() { return button.render() + " " + checkbox.render(); }
}

public class AbstractFactory {
    public static void main(String[] args) {
        System.out.println(new Screen(new LightWidgetFactory()).draw());
        System.out.println(new Screen(new DarkWidgetFactory()).draw());
    }
}
